# Pividori Laboratory onboarding information

# Mission statement

Our lab is dedicated to advancing precision medicine through the development and application of novel machine-learning methods.
Our primary focus is on developing novel algorithms and tools that integrate genetic studies with multi-omics data to unravel the complex interplay between genetics and disease.
We aim to pave the way for personalized diagnostic and therapeutic approaches to complex diseases by harnessing the power of computational analysis and systems biology.
We incorporate the latest advancements in machine learning and knowledge discovery to model the emergent complexity of biological systems with the aim of advancing our understanding of disease pathology and improving human health.

To accomplish this, we will write high-quality code and perform robust and reproducible analyses, ensuring our research is conducted within reproducible environments.
By adopting such practices, we aim to provide biologists who use our methods and web servers with high confidence in our results and processes.
Therefore, we are committed to maximizing openness and accessibility by making our source code available to the scientific community.

When disseminating our findings through publications and web servers, we will not only share the final results but also provide complete access to the reproducible environments that generated those results.
We recognize that trust is paramount, and to maintain that trust, we will present analytical code in which we can take pride.
As such, we will furnish reviewers and the scientific community with all the source code necessary to recreate the figures presented in our papers resulting from computational analyses.
By adopting a transparent and reproducible approach through reproducible environments and open-source code, we aim to foster a culture of trust, collaboration, and verifiability in the computational biology research field.

# Expectations

**Your role:** We expect that you will take primary responsibility for the success of your research project and career development.
As a member of the lab, you are expected to participate fully in the team.
In general, lab members are expected to follow weekday working hours that include 9:30 AM to 4:00 PM in their local timezone to facilitate discussion within the group.
We will aim to schedule meetings that respect these hours to the extent possible between the Eastern and Pacific time zones, recognizing that certain meetings (weekly kick-off and demo day) may fall outside of these hours in certain timezones.
In no cases should recurring meetings be scheduled that routinely fall outside of 8:00 AM to 5:00 PM for any lab member working within the continental United States.

**Milton's role:** Milton's goal is to facilitate your success as well as that of your project.
Within your project, Milton will serve as a sounding board for ideas, help you plan your project, and help devise experiments to test your hypotheses.
To facilitate your success, Milton will help you to plan your training, devise a career plan that can take you to where you want to go, advise you on your project-risk portfolio, and provide guidance on other elements of career and project development as needed.
Milton is by training a software engineer and as time allows, he will also provide feedback on your code if you request it.

**Deadlines:** Our lab aims to develop a reputation for high-quality science that is well-presented.
For this, abstracts for meetings must be shared with all co-authors, including Milton, at least one week prior to the deadline for submission.
Failure to abide by this guideline will result in missing whatever the opportunity in question is.

Trainees in the lab will often receive opportunities to present their work at scientific conferences.
These presentations reflect on the entire lab.
Oral presentations on projects must be presented to the research lab during a `braintrust` meeting, and lab members are expected to address feedback that is provided.
Once a project has been presented once and feedback incorporated, additional presentations are optional in advance of meetings.
Poster presentations should be shared in the `#general` slack channel at least a week before printing.

**Code of conduct:** All lab members and visitors are expected to agree with this code of conduct.
We will enforce this code.
We expect cooperation from all members to help ensure a safe environment for everybody.
The lab is dedicated to providing a harassment-free experience for everyone, regardless of gender, gender identity and expression, age, sexual orientation, disability, physical appearance, body size, race, or religion (or lack thereof).
We do not tolerate harassment of lab members in any form.
Sexual language and imagery are generally not appropriate for any lab venue, including lab meetings, presentations, or discussions.
However, do note that we work on biological matters, so work-related discussions of e.g., animal reproduction are appropriate.
Harassment includes offensive verbal comments related to gender, gender identity and expression, age, sexual orientation, disability, physical appearance, body size, race, religion, sexual images in public spaces, deliberate intimidation, stalking, following, harassing photography or recording, sustained disruption of talks or other events, inappropriate physical contact, and unwelcome sexual attention.
Members asked to stop any harassing behavior are expected to comply immediately.

If you are being harassed, notice that someone else is being harassed, or have any other concerns, please contact Milton Pividori immediately. <!-- If Milton is the cause of your concern, Dr. X (to.be.defined@university.edu) is a good informal point of contact; he/she does not work for Milton and has agreed to mediate. -->
For additional resolution paths, please see the University of Colorado Anschutz [ombuds](https://www.ucdenver.edu/offices/ombudsoffice) or [professionalism](https://www.cuanschutz.edu/offices/professionalism) offices.
The code of conduct section is licensed under a Creative Commons Attribution 3.0 Unported License. <http://2012.jsconf.us/#/about> & The Ada Initiative.
Please help by translating or improving: <http://github.com/leftlogic/confcodeofconduct.com>.

We expect members to follow these guidelines at any lab-related event.

**Authorship:** Our lab follows the [ICMJE's Uniform Requirements for Manuscripts Submitted to Biomedical Journals](https://www.icmje.org/recommendations/browse/roles-and-responsibilities/defining-the-role-of-authors-and-contributors.html) definitions of the roles of authors and contributors to our manuscripts.

**Ethics:** We expect lab members to be honest in scientific communications both within and outside the lab.
We expect that lab members will design experiments in a manner that minimizes both bias and self-deception.
We expect that lab members will keep agreements, be careful, and share their code and results openly with the scientific community.
We expect that credit will be given where credit is due, including in scientific writing.
Plagiarism is not tolerated.
While a full enumeration of ethical considerations is outside this document's scope, CU provides a [document](https://www.cu.edu/ope/aps/2027) that we recommend.

In addition, please don't hesitate to raise any questions or concerns that you have at any point with Milton.

**PhD student committees:** PhD students will interact with their qualifier and thesis committees.
Students should correspond with the coordinator for their graduate program to understand the expectations that exist around communication with committee members.
Questions about what document(s) their committee will expect to see and when they should be sent to the committee should be resolved with the coordinator at least a month in advance of a scheduled meeting.
Students in the Pividori Lab are not to provide food or drinks for committee members.
If the students are in a graduate program where a culture of providing food and drinks to committee members has developed, the students can include the information that no food or drink will be provided in an email in advance of the meeting and cite this policy.

**Conference travel:** We try to make sure that each member of the lab can travel to one conference per year of their choice outside of their home region.
The conference should be within the continental United States or cost-competitive with similar conferences in the continental US.
Lab members who travel to such a conference should submit an abstract for an oral presentation and poster and should present in whatever form is accepted at the meeting.
The conference should be topical for the lab member's research projects and the purpose must align with the grant(s) that support the lab member.
Lab members should first clear such travel with Milton.
Lab members who are invited to conferences or other presentation opportunities with their costs covered by the organization inviting them, e.g., as an invited speaker or keynote, are welcome to accept such invitations.
In all cases, conference travel should be noted on the lab attendance calendar (see below on how to use the lab's calendar).

# Communication

## General

**Slack:** We use Slack for rapid communication within the lab.
If you plan on sending an e-mail to someone within the lab, try a Slack message instead.
This helps to keep communications in one place, and Milton commits to respond to slacks (not necessarily immediately, but the same guarantee is not made for e-mail).
There are many channels on our lab's slack; however, it is recommended that newcomers join the following channels: `#general`, `#lab-meetings`, `#random`, `#wins`.

**HeyTaco:** We recognize that people regularly go above and beyond lab expectations.
We wanted a way to recognize each other when this happens.
We use HeyTaco.
This allows lab members to send a quick virtual thank you note and/or pat on the back.
If someone's paper gets accepted or someone helps you out with a programming question, congratulate or thank them.
Post a message that mentions any user in the `#wins` Slack channel, and they'll get a HeyTaco point.
When one member accumulates enough points, they take the lab out to lunch (Milton pays).

**Social media:** Lab members are encouraged to communicate through public social media.
If a lab member chooses to do so on an account that notes an affiliation with the lab, the lab member is expected to follow our code of conduct.
Certain employers may require a disclaimer that the lab member's views do not represent those of the employer.

**Projects:** By the nature of our research, lab members will often have the opportunity to participate in projects managed via private or publicly accessible source code repositories.
In these cases, lab members are expected to: follow the code of conduct, expect that private repositories will be world accessible, and communicate via the project-specific medium.
For example, when one lab member reports an issue on a project on GitHub, it would not be appropriate for another lab member to reply "I'll drop by your desk and show you how to solve that."
It would be most appropriate for the conversation to take place on GitHub issues.

**IP/Openness:** This is handled in accordance with the instructions from our research sponsors and university guidance.
Lab members must follow the [University of Colorado | Anschutz policies](https://www.cuanschutz.edu/cu-innovations/cu-innovations/innovator-resources/innovation-resource-guide/policies-documents) (for CU-affiliated lab members) and the agreements with our sponsors.
These often allow, encourage, or require openness.
If you have concerns at any point, set up a meeting with Milton to discuss these concerns.

**Space:** Space is assigned in accordance with the [Department of Biomedical Informatics Space Policies](https://olucdenver.sharepoint.com/sites/CenterforHealthAI939/Shared%20Documents/Forms/AllItems.aspx?csf=1&web=1&e=x1Oz36&cid=4490ca17%2D62a2%2D4093%2Da390%2Dcc8896df587d&FolderCTID=0x012000F2C178B56384E246B75AF7E31B7DEA98&id=%2Fsites%2FCenterforHealthAI939%2FShared%20Documents%2FGeneral%2FDBMI%20Final%20Policy%2FDBMI%20remote%2Dworking%20and%20space%20guidelines%5FFinal%2Epdf&parent=%2Fsites%2FCenterforHealthAI939%2FShared%20Documents%2FGeneral%2FDBMI%20Final%20Policy). For individuals meeting the criteria for assigned space, a lab member needs to fill out the [DBMI Space Request Form](https://app.smartsheet.com/b/form/575a2b48af774e3db5689745809c9a5c).

**Calendars:** When you join the lab, you should be added to the "Pividori Lab" group in Outlook.
In this group, we have a calendar that you should be able to see and modify.
In this calendar, we add:
1. *lab events*: mandatory events such as lab meetings, scrums, and group deadlines.
1. *attendance*: for noting individual availability (i.e. whether you'll be out of the office); it should be used, for example, to note vacations, conference travel, and other workday conflicts. For example, if you'll be out of the office for a few days, create an event in the "Pividori Lab" calendar with the title "Your name: out of office" or "Your name: attending ASHG."

**Accounts:** Lab members are expected to have accounts for the following and be members of the specified (organizations) if applicable:

+ GitHub (pivlab)
+ Outook (this is the CU account; you should be part of the "Pividori Lab" group)
+ Slack (Pividori Lab, pivlab.slack.com)

## Meetings

Our team's scrum process ("scrum" is a framework typically used in software development) involves three components:

1. A *weekly kick-off meeting* at 9:30 AM MT on Monday morning.
2. A *weekly demo day meeting* at 1:00 PM MT on Friday afternoon.
3. A *daily virtual scrum update*.

**Weekly kick-off meeting:** This is a quick (~5 minutes) virtual lab meeting where individuals will lay out their weekly goals on Zoom/Teams.
If you cannot join the kick-off meeting, you are expected to send a short message to the `#lab-meetings` channel (in Slack) with your update before the kick-off meeting.

**Weekly demo day meeting:** This is a quick (no more than 10 or 15 minutes) virtual lab meeting where team members show off an accomplishment from the week in 3 minutes or fewer.
This could be a new figure, a section of a paper, some code that they are particularly happy with, or something that we learned from a paper, poster, or research presentation.
For this, we use [PowerPoint slides](https://olucdenver-my.sharepoint.com/:p:/g/personal/milton_pividori_cuanschutz_edu/EYnxr99YOfREgFaPqe3_uEQB2hu8YFUQrLnqyuth3xKc1Q?e=PIL4hE) to share figures or paper sections for the demo day meeting; the link is pinned in the `#lab-meetings` channel on Slack and is not intended to be shared outside the lab.
If you are not already listed in the slides, feel free to add a slide with your name. Alternatively, screen sharing is possible for code demos or interactive weekly accomplishments.

**Daily virtual scrum update:**
The daily virtual scrum update should include an update from each individual to the [daily-tasks repository](https://github.com/pivlab/daily-tasks/).
An issue is automatically created for each workday, and they can be accessed [here](https://github.com/pivlab/daily-tasks/issues).
Each member of the lab is required to post a comment on the corresponding issue with a list of tasks planned for that day.

When writing this comment, consider the following:
1. what specific items you accomplished yesterday,
1. what specific items you plan to accomplish today,
1. who, if anyone, is blocking you,
1. who, if anyone, you are blocking.

**Lab meeting:** 
- Lab meetings are scheduled for one hour on Wednesdays. All members of the Pividori lab are expected to attend if possible. Meetings are expected to be a supportive environment for learning, constructive criticism, help, and scientific discussions. 

   **Meeting agenda:**  The format of each meeting will be chosen by the lab meeting lead from the options below. The lead will rotate among lab members (see below) within the group. Guests with aligned research interests may join with a supermajority vote of lab members (>2/3) and are expected to attend and participate fully.

    The lead will choose the format for each meeting. The different options for meeting formats are outlined below. Each member is expected to lead at least two Braintrust meetings per year (one every 6 months). 

   - **Format 1: Braintrust**
        - The meeting lead presents their own research/project to the group. Presenters often focus on open questions or challenges in their work. Occasionally, they present a new talk or set of slides that they intend to deliver at a meeting, job talk, etc. This is a way for the group to get familiar with each other's work. It is also a good way to get feedback, advice, or help with research if needed.
   - **Format 2: Tech talk/discussion**
        - Talks on commonly used tech in the labs, or strategies for staying on top of the literature, organization, etc.    
   - **Format 3: Post-conference presentations**
        - Journal club talk on favorite poster/talk. Either from each person or from a selected set that the group votes on.
   - **Format 4: Big ideas or projects**
        - This format is meant to help senior members practice for paper discussion sections/conclusions while helping newer members see where the boundaries of fields are.
   - **Format 5: Journal club**
       - Presentation to be given by meeting leader followed by group discussion. The meeting leader should aim to send the chosen paper one week before the scheduled meeting, and lab members are expected to be familiar with the content for discussion.
   - **Format 6: Preprint review**
       - Pre-print is discussed by the group. The discussion is led by the meeting leader,  and all members are expected to be familiar with the content. The review is written collaboratively, but another member (not the meeting leader) formats, formalizes and uploads to the pre-print server as a comment.

  **Lab meeting lead:**  A member of the Pividori lab is expected to present/lead each lab meeting as scheduled in the lab meeting calendar (sign up at [https://docs.google.com/document/d/1KeYg1_ijgHTMxQQLohcmtGX8wEYs_-ayUOn0kmtSq2U/edit?usp=sharing])
   - Each member is expected to sign up as lead by the end of the previous academic semester (e.g., sign up by December for the Spring semester, sign up by June for the Fall semester.)
   - Lab meeting sign-ups will happen twice a year. Sign-ups will open in December and June for the upcoming 6 months.
   - Each member will sign up as lead for at least 2 lab meetings each semester (6 months). At least one of these meetings is expected to be a Braintrust meeting.

  **Ad-hoc meetings:** After meetings have been scheduled for the semester, any member can add an ad hoc meeting as needed.
   - Ad hoc meetings are meant to help lab members get advice and help on projects, prepare for talks, oral exams, etc.

**Individual meetings:** We schedule weekly individual meetings.
 Once you join the lab, contact Milton to set up a time.
 These are set up for a term to accommodate class schedules.
 We don't reschedule these meetings by default if one of the parties (Milton or you) are out of town, so if you do want to meet in a week but travel conflicts, contact Milton to reschedule.
 The goal of the weekly meeting is to:

 1.  Discuss challenges.
 2.  Plan strategy (project related, personal career, etc).

**Triannual self reflection:** Every four months students, postdocs, and staff will individually meet with Milton to discuss their existing goals, current progress made and set goals for the next interval.
To prepare for these meetings students and staff are required to create an activity report that contains any of the following information (if applicable):

- publications: submitted/accepted/published
- grants/fellowships/scholarships: applied/awarded
- presentations delivered
- posters presented
- meeting abstracts: submitted/accepted
- software releases
- other honors
- goals for next session: What would you like to accomplish by the end of next cycle?
- self-reflection. What do you regard as your strengths and as areas where you need improvement?

The report should be in the form of a plain text file, markdown file, or PDF and the file should be called lastname-reflection-yearmonth (e.g. Pividori-reflection-202308.txt).
Submit the report in a direct message to Milton via slack.
<!-- How does this work at CU? - During the summer, graduate students are required to complete CU's individual development plan (IDP). -->
<!-- Post-docs must complete an IDP prior to their annual contract renewal. -->
<!-- This document covers more in-depth content than the regular triannual self reflection; therefore, the IDP can be used as a replacement annual report for that cycle. -->
<!-- Because much of the material is overlapping, trainees will benefit from preserving their self reflection materials in a format that supports copying and pasting to the IDP form. -->

# Source dode, data, and reproducibility

**Pride:** We expect lab members to sign their code, which means that source code contributions are attributable to an individual's account on GitHub.
To quote from _The Pragmatic Programmer_:

> Craftsmen of an earlier age were proud to sign their work.
You should be, tooâ€¦
People should see your name on a piece of code and expect it to be solid, well written, tested, and documented.

While some code will be proof-of-concept code, it should be of a form that inspires confidence.

**Language:** We write code for our analyses in Python or R, which allows everyone in the lab to know two languages and understand analytical code.
Code for visualization can be Python, R, or javascript.
Web server interface code uses javascript.

**Licensing:**
We release as many research outputs as possible under permissive open licenses.
This ensures lab research is reusable and reproducible, with minimal legal barriers.
The default license for software that should be applied to new lab related repositories is the [BSD-2-Clause Plus Patent License](https://spdx.org/licenses/BSD-2-Clause-Patent).
This license is [OSI-approved](https://opensource.org/licenses/BSDplusPatent) and [rated highly](https://blueoakcouncil.org/list) for its simplicity, compatability, and effectiveness.

In certain cases, a funding agency requires a different license or upstream restrictions require certain licensing.
In these cases, the lab may apply a different license.
If you have questions or concerns about licensing, feel free to raise them in Slack.

**Version control services:** Our primary version control service is GitHub, and we have a [`pivlab` account](https://github.com/pivlab) there.
We expect that lab members will maintain their code in repositories under these team accounts.
However, lab member should not commit to the branch that is shown as default on GitHub for any of these repositories.
Instead commits happen as described below to facilitate code review.

**Creating a Pividori Lab repository:**

1.  Create a repository under the team account.
2.  Immediately fork this repository into one that your user account owns.
3.  Make commits to your own repository, and move code back to the `pivlab` repository as described below.

**Getting code into Pividori Lab repositories:** Code moves from user repositories to `pivlab` repositories through a process of code review.
Code review is handled through pull requests.
The process is described briefly below.
Feel free to ask for guidance if you are uncomfortable with the process.
**We will revoke write access for failing to adhere to these rules.**

1.  Make changes to your code and commit them in your own repository first.
2.  Create a pull request into the repository owned by Pividori Lab.
3.  Name potential reviewers for your pull request.
4.  Once at least one lab member has approved your pull request, you or a reviewer may merge your pull request.
The only exception to this policy is this repository ("onboarding") where, in addition to the above rules, Milton must also approve the pull request.

**Composition of pull requests:** Each pull request may contain one or more changesets.
In keeping with good source control practice, each changeset or commit should contain *all* changes necessary for a particular fix or update.
In addition, each pull request should relate to no more than one functional area in the code base you are updating.
Keeping the pull request focused to one area makes it easier for your reviewers to provide thoughtful feedback.

**Reviewing pull requests:** We expect that all lab members will participate in review of pull requests.
If you get named by the submitter, it's courteous to review the request.
We have created a [checklist](docs/code_review_checklist.md) to facilitate review.
As a reviewer, you are responsible for making sure that all checklist guidelines are followed.

**Projects that didn't work:** We expect that repositories will contain failures (e.g. proof-of-concepts that didn't work).
This is ideal.
Being able to find them will make sure we don't make the same failure twice.

**Non-code versioning:** Non-code documents should be kept in a place that maintains version history.
The University of Colorado provides [OneDrive](https://medschool.zendesk.com/hc/en-us/sections/360005497433-OneDrive) for this purpose.

**Data management:** For publicly available data, scripts used to download and process these data should be preserved, as should the versions of items used in processing (e.g. probe to gene mappings).
These items should be version controlled.
Where possible, intermediate files of reasonable size can be stored to facilitate re-use, but the process to regenerate these files from publicly available data should be preserved.
When we generate data, they should be stored in a location where they are replicated and uploaded to the relevant database as soon as possible (e.g. GEO for gene expression, SRA for sequencing).

**Reproducibility:** We expect all lab members to maintain code that performs reproducible analyses.
This can be in the form of makefiles, shell scripts, or other automation approaches such as [pytask](https://pytask-dev.readthedocs.io/en/stable/index.html) that allow analyses to be automatically performed.
We expect that these scripts, including those to generate figures in papers generated as a consequence of such analyses, will be included in source control repositories (see "Getting Code into Pividori Lab Repositories") and made publicly available before or concurrent with the submission of preprint (if submitted) or manuscripts.
Combined with the review guidelines, this means that all code must have been reviewed for these documents to be submitted.

## How to modify this document

This is a living document.
The repository is at [GitHub](https://github.com/pivlab/onboarding).
To make changes, fork, edit the files you wish, and create a pull request.
The pull request process is handled as described in the Getting Code into Pividori Lab Repositories section of coding_and_software.

<!-- ## Additional Resources

+ [development tutorials](extras/development_tutorials.md)
+ [code review checklist](extras/code_review_checklist.md)
+ [deployment checklist](extras/deployment_checklist.md)
+ [infrastructure](extras/infrastructure.md)
+ [newcomers info](extras/newcomers_info.md) -->
